Verify country code on checkout

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Currently the checkout process doesn't validate if the posted country codes (in `billing_country` and `shipping_country` POST keys) actually exist, so it's possible to post a non-existing country code e.g. "XX" and it will be recorded as such in the billing or shipping address. This pull request adds validation for these keys in the checkout process.

Closes #27521.

### How to test the changes in this Pull Request:

1. Start with `master`.
2. Add the following at the end of `woocommerce.php` to simulate tampering with the form data:

```php
if ( array_key_exists( 'billing_country', $_POST ) ) $_POST['billing_country'] = 'XX';
if ( array_key_exists( 'shipping_country', $_POST ) ) $_POST['shipping_country'] = 'YY';
```

3. Place an order where the shipping address is the same as the billing address. The checkout will succeed, go check the order details and see that the country in the billing address is "XX".
4. Switch to this branch.
5. Repeat 3. Now the checkout should fail with a _'XX' is not a valid country code_ error.
6. Place an order where the shipping address is different from the billing address. Now there are two errors on checkout, one for 'XX' and another one for 'YY'.
7. Remove the hack on `woocommerce.php`, try to complete checkout without specifying a country and verify that the error message is still the proper one ("is a required field").

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - add validation of the posted country codes on checkout

Konamiman committed:
Validate checkout countries in the backend

Konamiman committed:
Fix code sniffer errors

vedanshujain commented on includes/class-wc-countries.php:
```diff
@@ -57,6 +57,24 @@ public function get_countries() {
+		$exists = array_key_exists( $country_code, $this->get_countries() );
+		if ( ! $exists && $throw ) {
+			/* translators: %s: ISO 3166-1 alpha-2 country code. */
+			throw new Exception( sprintf( __( "'%s' is not a valid country code", 'woocommerce' ), $country_code ) );
```
what do you think about moving the exception throwing logic to the parent code, to make the code there more directly readable?

Konamiman replied:
I was trying to provide a centralized point for exceptioning, but sure, readabilty is important. I've done the change.

vedanshujain reviewed (commented):
Looks good and testing well. Thanks for update to coding standards as well.

Left a minor comment, and can we also add unit tests for this?

Konamiman committed:
WC_Countries::country_exists no longer throws an exception.

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -825,12 +827,20 @@ protected function validate_posted_data( &$data, &$errors ) {
 		$this->check_cart_items();
 
-		if ( empty( $data['woocommerce_checkout_update_totals'] ) && empty( $data['terms'] ) && ! empty( $_POST['terms-field'] ) ) { // WPCS: input var ok, CSRF ok.
+		$billing_country = WC()->customer->get_shipping_country();
```
Looking at line https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-checkout.php#L759 it seems like this method is used as a fallback and not as default way to get country address (unless cart needs shipping). Perhaps we should move these checks to inside `validate_posted_data` method only if country is a required field (it might not be for some shops, I think looking at the linked code).

Konamiman replied:
I see an inconsistency here. On one hand the `validate_posted_data` method is doing

```php
$country = isset( $data[ $fieldset_key . '_country' ] ) ? $data[ $fieldset_key . '_country' ] : WC()->customer->{"get_{$fieldset_key}_country"}();
```

but on the other hand the `validate_checkout` method does just

```php
if ( WC()->cart->needs_shipping() ) {
	$shipping_country = WC()->customer->get_shipping_country();
```

so maybe that needs fixing too?
```

Konamiman replied:
After a second look I see that `WC()->customer->get_shipping_country()` is effectively getting the country from the request... not sure if that's intentional (it's confusing at least).

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -825,12 +827,20 @@ protected function validate_posted_data( &$data, &$errors ) {
+		$billing_country = WC()->customer->get_shipping_country();
+		if ( ! WC()->countries->country_exists( $billing_country ) ) {
+			/* translators: %s: ISO 3166-1 alpha-2 country code. */
+			throw new Exception( sprintf( __( "'%s' is not a valid country code", 'woocommerce' ), $billing_country ) );
```
instead of throwing immediately, should we be doing `$errors->add( $key . '_validation', ... );` to be consistent with other usages in this class?

vedanshujain reviewed (commented):
Thanks for updating, I noticed that country _might_ be optional in some cases, left couple of minor comments to that effect.
Also, can we add unit tests for this?

superdav42 commented on includes/class-wc-countries.php:
```diff
@@ -57,6 +57,17 @@ public function get_countries() {
+	 * @return bool True if the country is known to us, false otherwise.
+	 */
+	public function country_exists( $country_code ) {
+		return array_key_exists( $country_code, $this->get_countries() );
```
```suggestion
		return isset( $country_code, $this->get_countries() );
```
Since `isset` is a language construct it is slightly faster than calling the function `array_key_exists`. You only need to use array_key_exists if might be dealing with an array which contains null values. See https://www.php.net/manual/en/function.array-key-exists.php#example-5434

Konamiman replied:
Thanks for the suggestion but the actual syntax is slightly different: `isset( $this->get_countries()[ $country_code ] );` - I'll apply it and push by myself.

Konamiman committed:
Move country validation to 'validate_posted_data'.

Also, when shipping to billing address don't show the error twice,
and show the "we don't ship to X" error only if the country actually
exists (if it doesn't then we'll have shown the "X is not a valid
country code" already).

Konamiman committed:
Add unit tests for the WC_Checkout class.

Konamiman committed:
Use 'isset' instead of 'array_key_exists' for 'country_exists'.

Konamiman committed:
Fix in the logic of 'validate_checkout' regarding non-existing countries.

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -825,23 +834,27 @@ protected function validate_posted_data( &$data, &$errors ) {
 	 * @since  3.0.0
 	 * @param  array    $data   An array of posted data.
 	 * @param  WP_Error $errors Validation errors.
+	 * @throws Exception Invalid country code in billing or shipping address.
```
is this comment still correct?

Konamiman replied:
It's not! I'll remove it.

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -825,23 +834,27 @@ protected function validate_posted_data( &$data, &$errors ) {
-				/* translators: %s: shipping location */
-				$errors->add( 'shipping', sprintf( __( 'Unfortunately <strong>we do not ship %s</strong>. Please enter an alternative shipping address.', 'woocommerce' ), WC()->countries->shipping_to_prefix() . ' ' . WC()->customer->get_shipping_country() ) );
+			} elseif ( ! in_array( $shipping_country, array_keys( WC()->countries->get_shipping_countries() ), true ) ) {
+				if ( WC()->countries->country_exists( $shipping_country ) ) {
```
do we need this check here? Looks like we are already checking this in `validate_posted_data`

Konamiman replied:
I added this to prevent an "Unfortunately do not ship to (invalid country code)" notice, which would be pretty much useless once we have shown the "Invalid country code" one.

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -744,6 +746,13 @@ protected function validate_posted_data( &$data, &$errors ) {
 				$field_label = isset( $field['label'] ) ? $field['label'] : '';
 
+				if ( 'country' === $field['type'] &&
+					! ( 'shipping' === $fieldset_key && ! $data ['ship_to_different_address'] ) &&
```
should we add a clause for `$required` and `isset( $data[ $key ] )` incase country is not a required field at all. based on the code in callee functions, seems like if this field is not present, we can use `WC()->customer->get_shipping_country()` as a backup.

Konamiman replied:
But this code is part of the loop that iterates over the posted data, so it will validate the posted country only if there is any. Also there's a `if ( ! isset( $data[ $key ] ) ) continue;` at the beginning of the loop. Or you mean something else?

The `! ( 'shipping' === $fieldset_key && ! $data ['ship_to_different_address'] )` is there to prevent a duplicate "Invalid country code" error, since when "ship to different address" is unchecked then shipping and billing countries will be the same ones.

vedanshujain replied:
> But this code is part of the loop that iterates over the posted data

Ah sorry, I meant `empty` not `isset`.  Basically, if a field is not required then it can be empty as well right? Or not? We should probably bail from checking if the country is valid, if its not a required field and is empty. 

> The ! ( 'shipping' === $fieldset_key && ! $data ['ship_to_different_address'] ) is there to prevent a duplicate "Invalid country code" error, since when "ship to different address" is unchecked then shipping and billing countries will be the same ones.

Makes sense :+1:

Konamiman replied:
Ok, a couple of things:

1. Theres's a `$validate_fieldset` variable already for that, generated by calling `maybe_skip_fieldset`; so we can just use that.
2. `maybe_skip_fieldset` is already doing the `'shipping' === $fieldset_key && ! $data ['ship_to_different_address'] ` part so there's no need to repeat it.

So I've pushed a commit with these changes.

vedanshujain commented on includes/class-wc-checkout.php:
```diff
@@ -825,23 +834,27 @@ protected function validate_posted_data( &$data, &$errors ) {
 
 		if ( WC()->cart->needs_shipping() ) {
-			$shipping_country = WC()->customer->get_shipping_country();
+			$shipping_country = isset( $data['shipping_country'] ) ? $data['shipping_country'] : WC()->customer->get_shipping_country();
```
nice work to make this explicit 👍

vedanshujain reviewed (commented):
Overall looks good. I have left some minor comments, which are mostly questions for myself.

Also test suite looks pretty neat 💯, thanks for adding it!

Konamiman committed:
Remove no longer relevant exception doc comment

Konamiman committed:
validate_posted_data: check if we actually need to validate the shipping country.

vedanshujain reviewed (approved):
LGTM!
